src/decorators/with-context.decorator.spec.ts

```diff
@@ -0,0 +1,192 @@
+      );
+    });
+
+    it('should initialize context with custom initial context', () => {
+      const result = testClass.methodWithInitialContext();
+
+      expect(result).toHaveProperty('correlationId');
+      expect(result).toHaveProperty('userId', 'user_123');
+      expect(result).toHaveProperty('service', 'TestService');
+    });
+
+    it('should work with async methods', async () => {
+      const result = await testClass.asyncMethod();
+
+      expect(result).toHaveProperty('correlationId');
+      expect(typeof result.correlationId).toBe('string');
+
+      expect(result).toHaveProperty('asyncProcessing', 'processing');
+      expect(result).toHaveProperty('regularProcessing', 'processing');
+      expect(result).toHaveProperty('step', 'processing');
+    });
+  });
+
+  describe('Method Parameters', () => {
+    it('should preserve method parameters', () => {
+      const result = testClass.methodWithParameters('test', 42);
+
+      expect(result.params).toEqual({ param1: 'test', param2: 42 });
+      expect(result.context).toHaveProperty('correlationId');
+      expect(result.context).toHaveProperty('requestId', 'req_456');
+    });
+
+    it('should work with promise-returning methods', async () => {
+      const result = await testClass.methodThatReturnsPromise('test-value');
+
+      expect(result.value).toBe('test-value');
+      expect(result.context).toHaveProperty('correlationId');
+      expect(result.context).toHaveProperty('level', 'info');
+    });
+  });
+
+  describe('Context Updates', () => {
+    it('should allow context updates within the method', () => {
+      const result = testClass.methodThatUpdatesContext();
+
+      expect(result.initialContext).toHaveProperty('correlationId');
+      expect(result.initialContext).not.toHaveProperty('step');
+
+      expect(result.updatedContext).toHaveProperty('correlationId');
+      expect(result.updatedContext).toHaveProperty('step', 'processing');
+
+      // Should be the same correlationId
+      expect(result.initialContext.correlationId).toBe(
+        result.updatedContext.correlationId
+      );
+    });
+  });
+
+  describe('Context Isolation', () => {
+    it('should isolate context between different method calls', () => {
+      const result1 = testClass.simpleMethod();
+      const result2 = testClass.simpleMethod();
+
+      expect(result1.correlationId).not.toBe(result2.correlationId);
+    });
+
+    it('should not have context outside decorated methods', () => {
+      // Call decorated method to ensure context is created and cleaned up
+      testClass.simpleMethod();
+
+      // Context should be empty outside the decorated method
+      const contextOutside = ContextStore.getContext();
+      expect(Object.keys(contextOutside)).toHaveLength(0);
+    });
+  });
+
+  describe('Initial Context Override', () => {
+    it('should merge initial context with default correlationId', () => {
```

Comment:
what happens when you have two `WithContext` within the same request execution cycle? 
Foo -> WithContext{correlation_id: 1} -> Goo WithContext{correlation_id: 2}

let's have a test for it too so users would know what to expect